File: java/app/src/main/java/cribbage/internal/HandScorer.java
package cribbage.internal;

import com.google.common.collect.Lists;

import java.util.*;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.*;

public class HandScorer {

    public long scoreHand(Hand hand) {
        return scoreFifteens(hand) +
                scoreRuns(hand) +
                scorePairs(hand) +
                scoreFlushes(hand) +
                scoreNobs(hand);
    }

    long scoreFifteens(Hand hand) {
        final List<Card> allCards = hand.getAllCards();

        final List<List<Card>> allPossibleSetsOfCards =
                IntStream.range(1, allCards.size())
                        .boxed()
                        .flatMap(x -> selectNFromList(x, allCards).stream())
                        .collect(toList());

        final long totalNumberOfFifteens = allPossibleSetsOfCards.stream()
                .filter(this::isSumOfPointsFifteen)
                .count();

        return totalNumberOfFifteens * 2;
    }

    int scoreRuns(Hand hand) {
        final List<Card> allCards = hand.getAllCards();

        final List<List<Card>> allPossibleSetsOfCards =
                IntStream.rangeClosed(3, allCards.size())
                        .boxed()
                        .flatMap(x -> selectNFromList(x, allCards).stream())
                        .collect(toList());

        final List<List<Card>> allRuns = allPossibleSetsOfCards.stream()
                .filter(this::isConsecutive)
                .collect(toList());

        final OptionalInt longestPossibleRun = allRuns.stream()
                .mapToInt(List::size)
                .max();

        if (longestPossibleRun.isPresent()) {
            return longestPossibleRun.getAsInt();
        }
        else {
            return 0;
        }
    }

    long scorePairs(Hand hand) {
        final Map<Rank, List<Card>> groupedByRank = hand.getAllCards().stream().collect(groupingBy(Card::getRank));

        final long totalNumberOfTwoOfAKind = groupedByRank.values().stream()
                .filter(cards -> cards.size() == 2)
                .count();

        final long totalNumberOfThreeOfAKind = groupedByRank.values().stream()
                .filter(cards -> cards.size() == 3)
                .count();

        final long totalNumberOfFourOfAKind = groupedByRank.values().stream()
                .filter(cards -> cards.size() == 4)
                .count();

        return totalNumberOfTwoOfAKind * 2 +
                totalNumberOfThreeOfAKind * 6 +
                totalNumberOfFourOfAKind * 12;
    }

    long scoreFlushes(Hand hand) {
        // Count five card flushes
        final Map<Suit, List<Card>> allCardsBySuit =
                hand.getAllCards().stream().collect(groupingBy(Card::getSuit));

        final long numberOfFiveCardFlushes = allCardsBySuit.values().stream()
                .filter(cards -> cards.size() == 5)
                .count();

        // Count four card flushes
        final Map<Suit, List<Card>> handBySuit =
                hand.getCards().stream().collect(groupingBy(Card::getSuit));

        final long numberOfFourCardFlushes = handBySuit.values().stream()
                .filter(cards -> cards.size() == 4)
                .count();

        if (numberOfFiveCardFlushes == 1) {
            return 5;
        }
        else if (numberOfFourCardFlushes == 1) {
            return 4;
        }
        else {
            return 0;
        }
    }

    long scoreNobs(Hand hand) {
        final long totalNumberOfNobs = hand.getCards().stream()
                .filter(x -> x.getRank() == Rank.Jack && x.getSuit() == hand.getFaceUpCard().getSuit())
                .count();

        return totalNumberOfNobs;
    }

    private List<List<Card>> selectNFromList(int n, List<Card> cards) {
        final List<List<Card>> combinations = Lists.newArrayList();
        if (n == 0) {
            combinations.add(Lists.newArrayList());
        } else if (n == cards.size()) {
            combinations.add(cards);
        } else {

            final Card firstCard = cards.get(0);

            // Including first card
            final List<List<Card>> combinationsOfRest = selectNFromList(n - 1, cards.subList(1, cards.size()));
            for (List<Card> combination : combinationsOfRest) {
                final List<Card> withFirstCard = Lists.newArrayList(firstCard);
                withFirstCard.addAll(combination);
                combinations.add(withFirstCard);
            }

            // Not including first card
            for (List<Card> combination : selectNFromList(n, cards.subList(1, cards.size()))) {
                combinations.add(combination);
            }

        }
        return combinations;
    }

    private boolean isSumOfPointsFifteen(List<Card> cards) {
        final int total = cards.stream()
                .mapToInt(x -> x.getRank().pointsValue())
                .sum();

        return total == 15;
    }

    private boolean isConsecutive(List<Card> cards) {
        final Set<Rank> uniqueRanks = cards.stream()
                .map(Card::getRank)
                .collect(toSet());

        final List<Rank> sortedRanks = Lists.newArrayList(uniqueRanks);
        sortedRanks.sort(Comparator.comparing(Rank::asInt));

        return sortedRanks.size() == cards.size() &&
                sortedRanks.get(0).asInt() + cards.size() - 1 == sortedRanks.get(cards.size() - 1).asInt();
    }
}

File: scala/src/main/scala/cribbage/internal/HandScorer.scala
package cribbage.internal

import cribbage.internal.domain.{Card, Hand}

object HandScorer {

  def scoreHand(hand: Hand): Long = {
    scoreFifteens(hand) +
      scoreRuns(hand) + 
      scorePairs(hand) +
      scoreFlushes(hand) +
      scoreNobs(hand)
  }

  private[internal] def scoreFifteens(hand: Hand): Long = {
    val allPermutations =
      Range(1, hand.allCards.size)
        .flatMap(n => selectNFromList(n, hand.allCards))

    allPermutations.count(isSumOfPointsFifteen) * 2
  }

  private[internal] def scoreRuns(hand: Hand): Long = {
    val cardSets = Range.inclusive(3, hand.allCards.size)
      .flatMap(n => selectNFromList(n, hand.allCards))

    cardSets.filter(isRun).maxByOption(_.size) match {
      case Some(run) => run.size
      case None => 0
    }
  }

  private[internal] def scorePairs(hand: Hand): Long = {
    val groupedByRank = hand.allCards.groupBy(_.rank)

    val countNOfAKind = (n: Int) => groupedByRank.values.count(cards => cards.length == n)

    val pairs = countNOfAKind(2)
    val royalPairs = countNOfAKind(3)
    val doubleRoyalPairs = countNOfAKind(4)

    pairs * 2 +
      royalPairs * 6 +
      doubleRoyalPairs * 12
  }

  private[internal] def scoreFlushes(hand: Hand): Long = {
    val allCardsBySuit = hand.allCards.groupBy(_.suit)
    val hasFiveCardFlush = allCardsBySuit.values.exists(cards => cards.length == 5)

    val handCardsBySuit = hand.cards.groupBy(_.suit)
    val hasFourCardFlush = handCardsBySuit.values.exists(cards => cards.length == 4)

    if (hasFiveCardFlush) 5
    else if (hasFourCardFlush) 4
    else 0
  }
  
  private[internal] def scoreNobs(hand: Hand): Long = {
    hand.cards
      .count(card => card.rank == 11 && card.suit == hand.faceUpCard.suit)
  }

  private def selectNFromList[A](n: Int, items: List[A]): List[List[A]] = {
    if (n == 0) List(List())
    else if (n == items.length) List(items)
    else selectNFromList(n - 1, items.tail).map(items.head :: _) ++ selectNFromList(n, items.tail)
  }

  private def isSumOfPointsFifteen(cards: List[Card]): Boolean = {
    val points = (card: Card) => Math.min(card.rank, 10)
    cards.map(points).sum == 15
  }

  private def isRun(cards: List[Card]): Boolean = {
    val sortedByRank = cards.sortBy(_.rank)
    sortedByRank(cards.length - 1).rank - sortedByRank(0).rank == cards.length - 1
  }
}
